packages/dds/tree/src/feature-libraries/default-schema/schemaChecker.ts

```diff
@@ -0,0 +1,105 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	MapTree,
+	StoredSchemaCollection,
+	TreeFieldStoredSchema,
+	TreeNodeStoredSchema,
+	LeafNodeStoredSchema,
+	ObjectNodeStoredSchema,
+	MapNodeStoredSchema,
+} from "../../core/index.js";
+import { FullSchemaPolicy } from "../modular-schema/index.js";
+import { Multiplicity } from "../multiplicity.js";
+import { allowsValue } from "../valueUtilities.js";
+
+export function isNodeInSchema(
+	node: MapTree,
+	schema: TreeNodeStoredSchema,
+	nodeSchemaCollection: StoredSchemaCollection,
+	schemaPolicy: FullSchemaPolicy,
+): boolean {
+	if (
+		schema instanceof LeafNodeStoredSchema &&
+		(node.value === undefined ||
+			node.fields.size !== 0 ||
+			!allowsValue(schema.leafValue, node.value))
+	) {
+		return false;
+	}
+
+	if (schema instanceof ObjectNodeStoredSchema) {
+		if (node.fields.size !== schema.objectNodeFields.size) {
+			return false;
+		}
+		for (const [fieldKey, field] of node.fields) {
+			const fieldSchema = schema.objectNodeFields.get(fieldKey);
+			if (
+				fieldSchema === undefined ||
+				!isFieldInSchema(field, fieldSchema, nodeSchemaCollection, schemaPolicy)
+			) {
+				return false;
+			}
+		}
+	}
+
+	if (schema instanceof MapNodeStoredSchema) {
+		for (const field of node.fields.values()) {
+			if (!isFieldInSchema(field, schema.mapFields, nodeSchemaCollection, schemaPolicy)) {
+				return false;
+			}
+		}
+	}
+
+	return true;
+}
+
+export function isFieldInSchema(
+	field: MapTree[],
+	schema: TreeFieldStoredSchema,
+	nodeSchemaCollection: StoredSchemaCollection,
+	schemaPolicy: FullSchemaPolicy,
+): boolean {
+	const multiplicity = schemaPolicy.fieldKinds.get(schema.kind.identifier)?.multiplicity;
+	if (multiplicity === undefined || !complyWithMultiplicity(field.length, multiplicity)) {
+		return false;
+	}
+
+	if (schema.types !== undefined) {
+		for (const node of field) {
+			if (!schema.types.has(node.type)) {
+				return false;
+			}
+		}
+	}
+
+	for (const node of field) {
+		const nodeSchema = nodeSchemaCollection.nodeSchema.get(node.type);
+		if (
+			nodeSchema === undefined ||
+			!isNodeInSchema(node, nodeSchema, nodeSchemaCollection, schemaPolicy)
+		) {
+			return false;
+		}
+	}
```

Comment:
I think it would be better to merge these two loops, so we only make one pass over each node in the field.

That way we can factor out a functions that looks like:

```
export function isNodeUnionInSchema(
	node: MapTree,
	allowedTypes: TreeTypeSet,
	nodeSchemaCollection: StoredSchemaCollection,
	schemaPolicy: FullSchemaPolicy,
): boolean {
```

Having such a function would be useful, since we need it for cases like checking if a node is valid to be inserted into an sequence field. I picked the name "NodeUnion" to align with other places in our APIs where we deal with nodes which may be one of several allowed types, for example FlexTreeTypedNodeUnion.

Splitting the logic this way also does a bit better separation of concerns and should make more targeted unit testing of this part of the checking slightly easier.